tests/end2end/utils.py

```diff
@@ -664,3 +740,89 @@ def cleanup_and_import_example_certificate(cpc):
     }
 
     return console.certificates.import_certificate(cpc, props), props
+
+
+class StatusError(Exception):
+    """
+    Indicates that a desired status cannot be reached.
+    """
+    pass
+
+
+def ensure_lpar_inactive(lpar):
+    """
+    Ensure that the LPAR is inactive, regardless of what its current status is.
+
+    If this function returns, the status of the LPAR will be 'not-activated'.
+
+    If that status cannot be reached within the operation and status timeout
+    of the session, StatusError is raised.
+
+    Parameters:
+
+      lpar (zhmcclient.Lpar): The LPAR (must exist).
+
+    Raises:
+      StatusError: The inactive status cannot be reached.
+      zhmcclient.CeasedExistence: The LPAR no longer exists.
+      zhmcclient.Error: Any zhmcclient exception can happen, except
+        OperationTimeout and StatusTimeout (which result in StatusError).
+    """
+    org_status = pull_lpar_status(lpar)
+    if org_status == 'not-activated':
+        return
+
+    try:
+        lpar.deactivate(force=True)
+    except zhmcclient.OperationTimeout:
+        status = pull_lpar_status(lpar)
+        timeout = lpar.manager.session.retry_timeout_config.operation_timeout
+        raise StatusError(
+            "Could not get LPAR {lp!r} from status {os!r} into "
+            "status 'not-activated' within operation timeout {to}; "
+            "current status is: {s!r}".
+            format(lp=lpar.name, os=org_status, s=status, to=timeout))
+    except zhmcclient.StatusTimeout:
+        status = pull_lpar_status(lpar)
+        timeout = lpar.manager.session.retry_timeout_config.status_timeout
+        raise StatusError(
+            "Could not get LPAR {lp!r} from status {os!r} into "
+            "status 'not-activated' within status timeout {to}; "
+            "current status is: {s!r}".
+            format(lp=lpar.name, os=org_status, s=status, to=timeout))
+
+    # This is just an additional check which should not fail.
+    status = pull_lpar_status(lpar)
+    if status != 'not-activated':
+        raise StatusError(
+            "Could not get LPAR {lp!r} from status {os!r} into "
+            "status 'not-activated' for unknown reasons; "
+            "current status is: {s!r}".
+            format(lp=lpar.name, os=org_status, s=status))
+
+
+def pull_lpar_status(lpar):
+    """
+    Retrieve the current LPAR status on the HMC as fast as possible and return
+    it.
+
+    LPAR status values and their meaning:
+
+    Status         Resources allocated  OS running  Notes
+    ---------------------------------------------------------------------------
+    not-activated  no                   no
+    not-operating  yes                  no
+    operating      yes                  yes         No degradations
+    acceptable     yes                  yes         Some channel degradations
```

Comment:
There is still internal discussion going on about the value "acceptable".

Reply:
It seems that a value of "acceptable" is not shown at the WS-API. It is only shown in the SE GUI. I have updated the description accordingly. After this is finally clarified, we can move the description of status values into the user documentation, but for this internal test function this is good enough.